File: leetcode/py/p216-combination-sum-iii.py
# LeetCode 216. Combination Sum III.
from typing import List


class Solution:
    def backtrack(self, num: int, nums: List[int]):
        total = sum(nums)
        count = len(nums) + 1
        if count == self.k:
            last_num = self.n - total
            if last_num >= num and last_num < 10:
                self.output.append(nums + [last_num])
            return
        if total >= self.n:
            return
        for new_num in range(num, 9):
            self.backtrack(new_num + 1, nums + [new_num])

    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        assert k >= 2 and k <= 9
        assert n >= 1 and n <= 60
        self.k = k
        self.n = n
        self.output = []
        self.backtrack(1, [])
        return self.output


def test_combinationSum3():
    test_cases = (
        (2, 1, []),
        (2, 3, [[1, 2]]),
        (2, 4, [[1, 3]]),
        (2, 10, [[1, 9], [2, 8], [3, 7], [4, 6]]),
        (2, 17, [[8, 9]]),
        (3, 7, [[1, 2, 4]]),
        (3, 9, [[1, 2, 6], [1, 3, 5], [2, 3, 4]]),
        (4, 1, []),
        (9, 45, [[1, 2, 3, 4, 5, 6, 7, 8, 9]]),
        (9, 46, []),
        (8, 36, [[1, 2, 3, 4, 5, 6, 7, 8]]),
        (8, 37, [[1, 2, 3, 4, 5, 6, 7, 9]]),
        (8, 38, [[1, 2, 3, 4, 5, 6, 8, 9]]),
    )

    solution = Solution()
    for k, n, exp in test_cases:
        act = solution.combinationSum3(k, n)
        if exp != act:
            print(f"fail, k: {k}, n: {n}, exp: {exp}, act: {act}")


test_combinationSum3()

File: leetcode/cpp/p216-combination-sum-iii.cpp
// LeetCode 216. Combination Sum III.
#include "leetcode.hpp"

#include <cassert>
#include <iostream>
#include <vector>

class Solution {
    int _k { 0 };
    int _n { 0 };
    std::vector<int> _nums;
    std::vector<std::vector<int>> _output;
    const int MAX_K = 9;

public:
    void backtrack(int num, int total)
    {
        const auto count = static_cast<int>(_nums.size()) + 1;
        if (count == _k) {
            const auto last_num = _n - total;
            if (last_num >= num && last_num <= MAX_K) {
                _nums.push_back(last_num);
                _output.push_back(_nums);
                _nums.pop_back();
            }
            return;
        }
        if (total >= _n) {
            return;
        }
        for (int new_num = num; new_num <= MAX_K; ++new_num) {
            _nums.push_back(new_num);
            backtrack(new_num + 1, total + new_num);
            _nums.pop_back();
        }
    }
    std::vector<std::vector<int>> combinationSum3(int k, int n)
    {
        assert(k >= 2 && k <= MAX_K);
        assert(n >= 1 && n <= 60);
        _k = k;
        _n = n;
        _nums.clear();
        _output.clear();
        backtrack(1, 0);
        return _output;
    }
};

int main()
{
    struct TestCase {
        int k;
        int n;
        std::vector<std::vector<int>> exp;
    };

    const TestCase test_cases[] = {
        { 2, 1, {} },
        { 2, 3, { { 1, 2 } } },
        { 2, 4, { { 1, 3 } } },
        { 2, 10, { { 1, 9 }, { 2, 8 }, { 3, 7 }, { 4, 6 } } },
        { 2, 17, { { 8, 9 } } },
        { 3, 7, { { 1, 2, 4 } } },
        { 3, 9, { { 1, 2, 6 }, { 1, 3, 5 }, { 2, 3, 4 } } },
        { 4, 1, {} },
        { 9, 45, { { 1, 2, 3, 4, 5, 6, 7, 8, 9 } } },
        { 9, 46, {} },
        { 8, 36, { { 1, 2, 3, 4, 5, 6, 7, 8 } } },
        { 8, 37, { { 1, 2, 3, 4, 5, 6, 7, 9 } } },
        { 8, 38, { { 1, 2, 3, 4, 5, 6, 8, 9 } } },
    };
    Solution solution;
    for (const auto& tc : test_cases) {
        const auto act = solution.combinationSum3(tc.k, tc.n);
        if (tc.exp != act) {
            std::cerr << "Fail, k: " << tc.k
                      << ", n: " << tc.n
                      << ", exp: " << leetcode::to_string(tc.exp)
                      << ", act: " << leetcode::to_string(act)
                      << "\n";
        }
    }
}
